src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs

```diff
@@ -662,6 +679,49 @@ internal bool TryFindAnonymousType(AnonymousTypeManager.AnonymousTypeTemplateSym
                 return _anonymousTypeMap.TryGetValue(type.GetAnonymousTypeKey(), out otherType);
             }
 
+            internal bool TryFindAnonymousDelegate(AnonymousTypeManager.AnonymousDelegateTemplateSymbol type, out AnonymousTypeValue otherType)
+            {
+                Debug.Assert((object)type.ContainingSymbol == (object)_sourceAssembly.GlobalNamespace);
+
+                // Anonymous delegates are indexed by name, and the names are <>f__AnonymousDelegate0, 1, ... .
+                // Within a compilation, the index in the name is determined by source order, but the actual signature
+                // of the delegate type may change between generations. For instance, the signature of a lambda
+                // expression may have been changed explicitly, or another lambda expression may have been inserted
+                // ahead of this one in source order. Therefore, we need to verify the signatures of the delegate types
+                // are equivalent - by comparing the Invoke() method signatures.
+                if (_anonymousDelegateMap.TryGetValue(type.Name, out otherType) &&
+                    otherType.Type.GetInternalSymbol() is NamedTypeSymbol otherDelegateType &&
+                    isCorrespondingAnonymousDelegate(type, otherDelegateType))
+                {
+                    return true;
+                }
+
+                otherType = default;
+                return false;
+
+                bool isCorrespondingAnonymousDelegate(NamedTypeSymbol type, NamedTypeSymbol otherType)
+                {
+                    if (type.Arity != otherType.Arity)
+                    {
+                        return false;
+                    }
+
+                    type = SubstituteTypeParameters(type);
+                    otherType = SubstituteTypeParameters(otherType);
+
+                    return type.DelegateInvokeMethod is { } invokeMethod &&
+                        otherType.DelegateInvokeMethod is { } otherInvokeMethod &&
+                        invokeMethod.Parameters.SequenceEqual(otherInvokeMethod.Parameters, (x, y) => isCorrespondingType(x.Type, y.Type)) &&
+                        isCorrespondingType(invokeMethod.ReturnType, otherInvokeMethod.ReturnType);
+                }
+
+                bool isCorrespondingType(TypeSymbol type, TypeSymbol expectedType)
+                {
+                    return this.Visit(type) is TypeSymbol otherType &&
+                        otherType.Equals(expectedType, TypeCompareKind.AllIgnoreOptions);
```

Comment:
>AllIgnoreOptions [](http://example.com/codeflow?start=71&length=16)

Should we pay attention to custom modifiers? Why not?

Reply:
The two assemblies are two versions of the same source assembly, so it's not clear custom modifiers would affect this.

Reply:
> The two assemblies are two versions of the same source assembly, so it's not clear custom modifiers would affect this.

If here our goal is metadata level equality, then I think the safest way to compare is to take into account everything that affects signatures in metadata (```TypeCompareKind.CLRSignatureCompareOptions```). Also, we probably should compare ```TypeWithAnnotations``` to avoid ignoring "interesting" things.